### PR TITLE
test: Add check_coverage method in testlib.py

### PR DESCRIPTION
**Pull request by henrywang**
check_coverage method collects code coverage result and saves it into file.

### CODE REVIEW

File: test/common/testlib.py [hunk 1]
```diff
@@ -1305,6 +1307,23 @@ def write_file(self, path, content):
         if self.is_nondestructive():
             self.addCleanup(m.execute, "rm -f {0}".format(path))
 
+    def check_coverage(self):
+        """Collect code coverage result and save to file
+
+        Hash result as part of file name, like "coverage-<coverage hash result>.json",
+        to avoid duplicate coverage file and save file to ".nyc_output" folder.
+        """
+        coverage_data = self.browser.eval_js("window.__coverage__", no_trace=True)
```
Comment (marusak) [inline test/common/testlib.py:1316]: `window.__coverage__` does not exist in firefox nor chrome.
Where does this come from? Is it some plugin in your browser? If so, then tests need to be able to install/load this plugin into browser.

File: test/common/testlib.py [hunk 2]
```diff
@@ -1305,6 +1307,23 @@ def write_file(self, path, content):
...
+        str_coverage = json.dumps(coverage_data)
+        hash_str = hashlib.sha256(str_coverage.encode('utf-8')).hexdigest()
+
+        cov_out_dir = pathlib.Path(".nyc_output")
+        cov_out_dir.mkdir(exist_ok=True)
+
+        cov_out_file = cov_out_dir / "coverage-{}.json".format(hash_str)
+        # avoid duplecate result file
+        if not cov_out_file.is_file():
+            cov_out_file.write_text(str_coverage)
```
Comment (marusak) [inline test/common/testlib.py:1326]: we collect axe report in very similar fashion (check `check_axe` in the same file) and we do:
```
        filename = "{0}{1}-axe.json.gz".format(label or self.label(), suffix)   
        with gzip.open(filename, "wb") as f:                                    
            f.write(json.dumps(report).encode('UTF-8'))                         
        print("Wrote accessibility report to " + filename)                      
        attach(filename)  
```
and I think you would need to do something similar (like gzip, print that there was something saved and `attach` it, so it is actually available in CI)

File: test/common/testlib.py (hunk 1 above)
Comment (henrywang) [reply test/common/testlib.py:1316]: This's based on UI test code coverage tool [istanbul](https://istanbul.js.org/), which needs a babel plugin [babel-plugin-istanbul](https://github.com/istanbuljs/babel-plugin-istanbul). The plugin [instruments code](https://github.com/weldr/cockpit-composer/blob/7f295a8ed5da2dd91f3f0328f480dc6beeb3e72c/webpack.config.js#L19) with Istanbul coverage. c-composer did this [here](https://github.com/weldr/cockpit-composer/blob/7f295a8ed5da2dd91f3f0328f480dc6beeb3e72c/Makefile#L9).
And cockpit ci container has the file (~/.config/codecov-token) already.

File: test/common/testlib.py (hunk 2 above)
Comment (henrywang) [reply test/common/testlib.py:1326]: It's not a report, it's used by the code [here](https://github.com/weldr/cockpit-composer/blob/7f295a8ed5da2dd91f3f0328f480dc6beeb3e72c/test/check-application#L166). The report can be viewed from [codecov.io](https://codecov.io/gh/weldr/cockpit-composer). So we do not need these coverage files locally.

**Comment by mvollmer** [comment]
FYI, I started on this a while ago: https://github.com/cockpit-project/cockpit/pull/11043

**Comment by henrywang** [comment]
> FYI, I started on this a while ago: #11043
I'm switching cockpit-composer test frame work to CDP and "unittest", cockpit is using right now. So I have to add this into testlib now for convenience!

**Comment by henrywang** [comment]
Close this PR because it's included in a subclass of `MachineCase` in PR [#931](https://github.com/osbuild/cockpit-composer/pull/931) in cockpit-composer test.

**Pull request closed without merging**